### PR TITLE
Sweeps simulation Snakefile

### PR DESCRIPTION
**Pull request by mufernando**
WIP

### CODE REVIEW

**Comment by mufernando** [comment]
- [] fix duplication of neutral sims due to annot

**Comment by nspope** [comment]
This is working with:
```
snakemake -c1 --snakefile workflows/sweep_simulate.snake
```
But, I'm not sure how to integrate this into the main `Snakefile` without screwing stuff up -- as this workflow uses a different configfile from the DFE stuff. Do we define separate configfiles for each module in `Snakefile`? @andrewkern do you have a suggestion?

Also, I changed things so we are simulating (possibly overlapping) windows instead of the entire chromosome. For now, this'll let us expand out the simulation grid without a huge computational burden. Ultimately the plan is to simulate entire chromosomes but  "trim" to the windows -- I'll add a toggle to the config for this behavior. (though this might only be possible in practice with a smaller chromosome)

**Comment by andrewkern** [comment]
yeah you could definitely define a separate config file here and in this snakefile point to it

**Comment by andrewkern** [comment]
frankly the config stuff could use a bit of an overhaul -- last I looked at it there was a lot of hard coded redundant paths

**Comment by nspope** [comment]
Cool, thanks -- I'll avoid adding more stuff to `Snakefile` for the time being, then.

**Comment by mufernando** [comment]
I simulated chromosome 22 with the Gamma_K17 DFE on exons and no scaling. It took me 1h18min on sesame.

```
time stdpopsim -vv -e slim --slim-scaling-factor 1 HomSap -d OutOfAfrica_3G09 YRI:10 --dfe Gamma_K17 --dfe-annotation ensembl_havana_104_CDS -c chr22 -o foo.ts &>log.txt
stdpopsim -vv -e slim --slim-scaling-factor 1 HomSap -d OutOfAfrica_3G09       4689.52s user 220.01s system 100% cpu 1:21:42.95 total_```

**Comment by andrewkern** [comment]
wow, very nice. using the cluster we could definitely get 100s of reps for full chr22.

**Comment by nspope** [comment]
@andrewkern yeah, but if we're also varying sweep location across a fine grid that'll be prohibitively costly, I think. @mufernando just benchmarked further and it takes ~16 minutes to simulate 20% of chr22 with no scaling, and 1.5min with a scaling factor of 4. So I think windowing + scaling is the way to go, as long as we don't do it too aggressively.

**Comment by andrewkern** [comment]
i agree for a fine grid, but i think its perhaps worth our time to take one or a small subset of locations and ask, "does full chrom simulation matter much?"

essentially we'd be asking about chrom-wide effects of BGS

**Comment by nspope** [comment]
yeah I agree! e.g., lay down a sweep in the middle of chr22, and simulate the entire chrom and progressively smaller windows centered around the sweep. If there are boundary effects they'd be visible in patterns of diversity at the window edges, I guess?

**Comment by andrewkern** [comment]
btw I started drafting the diploshic module for this part of the analysis [here](https://github.com/andrewkern/analysis2/tree/diploshic)

**Comment by nspope** [comment]
Looks great, thanks Murillo! I'm assuming we'll be saving sims as vcf to be fed to the sweep detection methods, so I think we should also add a routine to dump vcf (for the focal window only) alongside the tree sequences.

File: conda-requirements.txt [hunk 1]
```diff
@@ -14,4 +14,4 @@ blas
 gsl
 scikit-learn
 dadi
-slim=3.7.1
\ No newline at end of file
+slim=4.0
```
Comment (andrewkern) [inline conda-requirements.txt:17]: worried about this move to 4.0-- is the "tick" timescale going to mess anything up? probably not as we are just doing WF models, yes?

File: workflows/sweep_simulate.snake [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+                                ts.metadata['windowing']['window_bright']-offset], dtype=np.int32)
+    # removing duplicates if bleft and left are the same or brighta and right
+    windows_bpoints = np.unique(windows_bpoints)
+    # something wonky happening with float conversion
+    if windows_bpoints[-1] == ts.sequence_length - 1:
+        windows_bpoints[-1] = ts.sequence_length
+    assert (windows_bpoints.shape == (2,)) or windows_bpoints.shape == (4,)
+    taj=ts.Tajimas_D(windows=windows_bpoints,mode="site")
+    focal_taj = taj[len(taj)//2]
+    return focal_taj, taj
```
Comment (andrewkern) [inline workflows/sweep_simulate.snake:138]: i think these function should be put in a module, perhaps named `sweep_utils.py`. it's useful to separate the code from the workflow.

File: workflows/sweep_simulate.snake [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+        for suffix in [".trees", ".stats.npz"] 
+        ] + [output_dir +  f"/simulated_data/sweeps/sweep/{annot}/{dfe}/{coeff}/{time}/{seed}/sim_{chrom}_{left:.0f}_{right:.0f}{suffix}"
+        for left, right in windows
+        for seed in seed_array
+        for dfe in dfe_list
+        for annot in annotation_list
+        for coeff in sel_coeffs
+        for time in times
+        for suffix in [".trees", ".stats.npz"] 
+        ]
```
Comment (andrewkern) [inline workflows/sweep_simulate.snake:165]: ... this list comprehension... 👀

File: workflows/sweep_simulate.snake [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+        output_dir + "/simulated_data/sweeps/boundary_sims/sim_{seed}_{region_size}.trees"
+    resources: time=6000, mem_mb=100000
+    run:
+        model = species.get_demographic_model(demo_model["id"])
+        mut_rate = model.mutation_rate 
+        # Use new sample specification, give selected pop and sample size as config arguments
+        samples = demo_model["samples"]
+        dfe = species.get_dfe(dfe_list[0])
+        # reducing number of del muts bc it's too much!
+        dfe.proportions = [1-(0.7*0.25), 0.7*0.25]
```
Comment (andrewkern) [inline workflows/sweep_simulate.snake:181]: looks like you are hard coding the DFE? what's up?

File: workflows/sweep_simulate.snake [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+            model,
+            contig,
+            samples,
+            slim_scaling_factor=config["slim_scaling_factor"],
+            slim_burn_in=config["slim_burn_in"],
+            seed=int(wildcards.seed),
+        )
+        ts.dump(output[0])
+
+rule neutral:
```
Comment (andrewkern) [inline workflows/sweep_simulate.snake:197]: do we have to do these simulations? can we instead use the simulations that are already being produced by the rest of the pipeline for neutral sims?

File: workflows/sweep_simulate.snake [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+            seed=int(wildcards.seed),
+        )
+        ts = add_window_metadata(ts, window_left, window_right, window_left, window_right)
+        ts.dump(output[0])
+
+rule bgs:
+    input:
+    output:
+        output_dir + f"/simulated_data/sweeps/bgs/{{annot}}/{{dfe}}/{{seed}}/sim_{chrom}_{{left}}_{{right}}.trees", 
+    resources: time=3000, mem_mb=4000
```
Comment (andrewkern) [inline workflows/sweep_simulate.snake:232]: these resources seem specific -- was this tuned?

**Review by andrewkern — COMMENTED** [review]
looks pretty clean. my biggest comment is to pull out the python functions you've defined here to their own module, just to keep the functions separate from the workflow

File: conda-requirements.txt (hunk 1 above)
Comment (petrelharp) [reply conda-requirements.txt:17]: stdpopsim itself has moved to slim v4, so we better have this!

File: workflows/sweep_simulate.snake (hunk 3 above)
Comment (mufernando) [reply workflows/sweep_simulate.snake]: I made these a bit more readable!

File: workflows/sweep_simulate.snake (hunk 5 above)
Comment (mufernando) [reply workflows/sweep_simulate.snake:209]: What do you mean?

File: workflows/sweep_simulate.snake (hunk 6 above)
Comment (mufernando) [reply workflows/sweep_simulate.snake:244]: the memory was tuned a bit, the time was not.

File: workflows/sweep_simulate.snake (hunk 5 above)
Comment (andrewkern) [reply workflows/sweep_simulate.snake:209]: i mean that we might be able to grab simulations from the `n_t.snake` workflow?

**Comment by andrewkern** [comment]
@mufernando do we need `results/simulated_data/sweeps/boundary_effect_bgs.png` in the PR? seems like just an image to post on an issue?

**Comment by andrewkern** [comment]
also this is probably ready to come out of draft mode?

**Comment by mufernando** [comment]
@andrewkern I thought that figure would make it to the supp material at some point?

**Comment by andrewkern** [comment]
okay sounds good

File: workflows/sweep_simulate.snake [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+    sample_sets = [ts.samples(population=all_pops_in_ts.index(pop)) for pop in target_pops]
+    fh = open(output[0], 'w')
+    focal_taj, _ = compute_stats(ts, sample_sets)
+    assert focal_taj.shape == (len(target_pops),)
+    print(f'{header}', file=fh, flush=True)
+    for t,p in zip(focal_taj,target_pops):
+        print(f'{meta_str}\t{p}\t{t}', file=fh, flush=True)
+    fh.close()
+
+rule get_stats:
```
Comment (andrewkern) [inline workflows/sweep_simulate.snake:364]: in a subsequent PR I'm hoping to add a rule that would use `scikit-allel` to calculate stats like haplotype homozygosity. maybe change this rule name to `get_stats_tskit`?

File: workflows/sweep_simulate.snake [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+            slim_scaling_factor=config["slim_scaling_factor"],
+            slim_burn_in=config["slim_burn_in"],
+            seed=int(wildcards.seed),
+        )
+        ts = add_window_metadata(ts, window_left, window_right, b_window_left, b_window_right, extra=f"sweep_time={time};")
+        ts.dump(output[0])
+
+def _get_stats(input, output, params_dict, target_pops):
+    meta = ["model","demo","seed","chrom","left","right","annot","dfe","coeff"]
+    stats = ["taj"]
```
Comment (andrewkern) [inline workflows/sweep_simulate.snake:348]: would be good to get a windowed look at \pi in the focal window just to confirm that we are parameterizing sims in a way that allows us to see a hitchhiking effect

Comment (andrewkern) [reply workflows/sweep_simulate.snake:348]: in looking at mean TajD over a small number of replicates it looks like we may need to tune the sweep params a bit 
image

**Review by andrewkern — COMMENTED** [review]
this is working 🎉 !!

but it looks like we have to tune the sweep params a bit to make sure we are laying down a detectable sweep(s) that will impact genomic variation in our simulations.

File: workflows/sweep_simulate.snake [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+            samples,
+            extended_events=extended_events,
+            slim_scaling_factor=config["slim_scaling_factor"],
+            slim_burn_in=config["slim_burn_in"],
+            seed=int(wildcards.seed),
+        )
+        ts = add_window_metadata(ts, window_left, window_right, b_window_left, b_window_right)
+        ts.dump(output[0])
+
+rule sweep:
```
[the remark this answers is not in the document]
Comment (nspope) [reply workflows/sweep_simulate.snake:287]: This has been popping up for a while during recap/rescale in stdpopsim -- it doesn't indicate anything is broken AFAICT (although we should figure out a way to suppress it)

File: workflows/sweep_simulate.snake [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+        gamma = coal_T* coeff #2Ns
+        T_f = 4 * (np.log(gamma) + 0.5772 - (1/gamma)) / coeff # this is from Charlesworth 2020/Pennings and Hermisson 2007
+        time = np.floor(1.5*T_f) # I dilated the time a bit because of variance (don't want to lose too many sims due to this -- note we are conditioning on sweep_min_freq)
+        print(coal_T, gamma, T_f, time)
+        contig.add_single_site("sweep", coordinate=coord)
+        extended_events = stdpopsim.ext.selective_sweep(
+            "sweep",
+            population=wildcards.popu, #selected pop
+            selection_coeff=coeff,
+            mutation_generation_ago=time,
```
[the remark this answers is not in the document]
Comment (nspope) [reply workflows/sweep_simulate.snake:328]: ultimately yes-- but I think we should keep as much fixed as possible for the moment, while we're making sure this all works

File: workflows/sweep_simulate.snake [hunk 11]
```diff
@@ -0,0 +1,432 @@
...
+
+        # sweep parameters
+        coord = ((window_left+window_right)//2) # the API does the right thing even if it is chunked
+        coeff = float(wildcards.coeff)
+        # Defining how long ago beneficial mutations arise
+        # First getting 2N from the model and pop
+        debug = model.model.debug()
+        coal_T = debug.mean_coalescence_time({wildcards.popu: 2})
+        gamma = coal_T* coeff #2Ns
+        T_f = 4 * (np.log(gamma) + 0.5772 - (1/gamma)) / coeff # this is from Charlesworth 2020/Pennings and Hermisson 2007
```
Comment (andrewkern) [inline workflows/sweep_simulate.snake:320]: i think the coefficient here may be off by a factor of 2?

**Comment by nspope** [comment]
@mufernando do you mind if I push some commits with helper functions for sweepfinder? Or should I hold off for now

**Comment by mufernando** [comment]
I think we should merge this and start separate PRs

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by andrewkern** [comment]
@mufernando do you want to make any more edits to this at this point?

**Pull request merged**